### PR TITLE
[lldb][RISCV] Support optionally disabled FPR for riscv64

### PR DESCRIPTION
**Pull request by AlexeyMerzlyakov**
The PR adds the support optionally enabled/disabled FP-registers to LLDB `RegisterInfoPOSIX_riscv64`. This situation might take place for RISC-V builds having no FP-registers, like RV64IMAC or RV64IMACV.

To aim this, patch adds `opt_regsets` flags mechanism. It re-works RegisterInfo class to work with flexibly allocated (depending on  `opt_regsets` flag) `m_register_sets` and `m_register_infos` vectors instead of statically defined structures. The registration of regsets is being arranged by `m_per_regset_regnum_range` map.

The patch flows are spread to `NativeRegisterContextLinux_riscv64` and `RegisterContextCorePOSIX_riscv64` classes, that were tested on:
 - x86_64 host working with coredumps
 - RV64GC and RV64IMAC targets working with coredumps and natively in run-time with binaries

`EmulateInstructionRISCV` is out of scope of this patch, and its behavior did not change, using maximum set of registers.

According testcase built for RV64IMAC (no-FPR) was added to `TestLinuxCore.py`.

### CODE REVIEW

**Comment by AlexeyMerzlyakov** [comment]
The change - is  a result of https://github.com/llvm/llvm-project/pull/93297 discussions about FP-register class refactoring

**Comment by DavidSpickett** [comment]
Can you remind me what vpr is as opposed to fpr, is that V as in Vector?

File: lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py
Comment (DavidSpickett) [inline lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:775]: If everything is optional in risc-v, it may be better to label these tests by what they contain not what they don't.

riscv64_gpr_only
riscv64_gpr_fpr

**Review by DavidSpickett — COMMENTED** [review]
Using the "dynamic" (as aarch64 calls it) regset idea is certainly the way to go. AArch64 has a bit of a strange setup where we have 2 fixed layouts, then everything else is optional. Possibly it needs to be this way because of SVE changing length, I never dug into it. These days I just use the optional mechanisms for everything.

**Comment by DavidSpickett** [comment]
Also consider adding "for riscv64" to the PR title, so it is clear to readers that 32 bit is not included.

File: lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py
Comment (DavidSpickett) [reply lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:775]: AArch64 has been saved from this problem by a lot of the new register state being weird enough it doesn't fit in this test very well. So for example scalable matrix extension ended up in `lldb/test/API/linux/aarch64/sme_core_file/TestAArch64LinuxSMECoreFile.py`.

But adding basics like FP here in this file is fine.

**Comment by francisvm** [comment]
Why is this riscv64-only?

**Comment by DavidSpickett** [comment]
Just from lldb's side, the code is organised such that we end up with separate code for 32 and 64 bit in a lot of cases. So a PR for 64 bit only is fine (in fact preferable, as this PR will provide a template for anyone wanting to implement 32 bit support).

Of course it's up to @AlexeyMerzlyakov what their plans are here.

**Comment by AlexeyMerzlyakov** [comment]
@DavidSpickett, thank you for the review and sorry for late reply, I've just returned from the BT w/o an access to my main PC.

Yes, the way to handle the register sets is pretty similar that ARM64 is using in dynamic model (if I understand it correctly), so all further added register sets could use this as a baseline, I believe.
Regarding review comments, I'll meet them shortly.

> Can you remind me what vpr is as opposed to fpr, is that V as in Vector?

Yes, everything is all right. The main thing here - is that some RISC-V platforms might have only general-purpose and vector registers, but non of floating-point ones (e.g. for the case of targets say specifically shaped for ML/AI). For this case, VPR will be placed in totaling structure right after GPR-s. Here it is important to fill registers arrays and numbers (offsets) correctly.

> Of course it's up to @AlexeyMerzlyakov what their plans are here.

Unfortunately, I have no plans on RV32-s right now. However, the 32-bit support could be added as a set of separate plugins for this target architecture using RV64 as a reference.

**Comment by DavidSpickett** [comment]
> The main thing here - is that some RISC-V platforms might have only general-purpose and vector registers, but non of floating-point ones (e.g. for the case of targets say specifically shaped for ML/AI). 

I had not considered that. I'm used to Arm where if you want to have the new vector extensions you have to have the original FP (Neon) extension too.

Even more reason to make it all dynamic.

File: lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py
Comment (AlexeyMerzlyakov) [reply lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:775]: Just to be on the same page:
Should I move `test_riscv64_regs_gpr`/`test_riscv64_regs_gpr_fpr` tests containing register value checks to separate place (like `lldb/test/API/linux/aarch64/sme_core_file/TestAArch64LinuxSMECoreFile.py` or better to `lldb/test/API/linux/riscv64/fpr_core_file/TestRV64FPRCoreFile.py`). And will keep main tests for RISC-V cores (`test_riscv64_gpr` and `test_riscv64_gpr_fpr`) in `TestLinuxCore.py`, as they are using `do_test()` infrastructure from this file? Or you mean something else?

Comment (DavidSpickett) [reply lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:775]: I'm saying that if risc-v grows a lot of strange register state, it will be better served by separate test cases.

For this PR, just consider renaming the test functions here so that they list what they are testing. So riscv64_gpr, as opposed to to riscv64_nofpr.

Comment (DavidSpickett) [reply lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:775]: As an example of what I mean about future extensions being "too weird" -

In the new Arm extension support I'm doing right now, I wrote a program file that can be used for live processes, and to generate a core file. This keeps all tests for that extension in one place. This helps because I have to use weird assembly to trigger these things, and I only have to write that once. And I don't have to put that in the standard corefile tests with some #ifdefs which get confusing.

With floating point, you can write normal C code to set values, which can be included easily in a program file that's quite generic.

Comment (DavidSpickett) [reply lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:775]: > For this PR, just consider renaming the test functions here so that they list what they are testing. So riscv64_gpr, as opposed to to riscv64_nofpr.

Or if you really want to be super clear: `riscv_gpr_only`

File: lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py
Comment (AlexeyMerzlyakov) [reply lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:777]: Re-generated executable and coredump for `*_fpr` test-cases using `main_fpr.c` instead of `main.c`, which has non-zero `fa5` FP-register during the crash. Adjusted test-cases accordingly in https://github.com/llvm/llvm-project/pull/104547/commits/e656524b63eed3a17e132e2280d44bf0e70e3c68

**Comment by AlexeyMerzlyakov** [comment]
Forse-pushed the branch to resolve merge-conflicts after no-update for a some time

File: lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py
Comment (DavidSpickett) [inline lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:731]: I would replace this line with:
```
# fa5 is non-zero and checked in the list above.
```

**Comment by DavidSpickett** [comment]
Just one more thing then this is good to go.

Do you have commit access or will you need me to merge it for you?

File: lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py
Comment (AlexeyMerzlyakov) [reply lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:731]: Done in https://github.com/llvm/llvm-project/pull/104547/commits/0379569846c717bcc3a0e6ecfb1c7aec4a3f218e

**Comment by AlexeyMerzlyakov** [comment]
> Do you have commit access or will you need me to merge it for you?

No, I have not. Could you please proceed this? Thanks

**Review by DavidSpickett — APPROVED** [review]
Thanks!

**Comment by DavidSpickett** [comment]
Hi @AlexeyMerzlyakov I am looking at commits that warrant a release note for LLDB 20, and I think this is one of them. If you want to do that yourself you can send a PR adding a bullet point to https://github.com/llvm/llvm-project/blob/main/llvm/docs/ReleaseNotes.md#changes-to-lldb, describing the change.

If you do not want to, or do not have the time to, I can write it myself. If I do not hear anything from you, I'll assume that I can go ahead with that.

**Comment by AlexeyMerzlyakov** [comment]
Hi, @DavidSpickett. Sure, no problem, I'll make an update PR for ReleaseNotes with optionally enabled/disabled register sets soon.

**Pull request merged**
